### PR TITLE
move etcd client encapsulation from proxy to gost

### PR DESCRIPTION
**Pull request by ztelur**
**What this PR does**:

move etcd client encapsulation to gost

**Which issue(s) this PR fixes**:

Fixes #

- add etcd client encapsulation

### CODE REVIEW

**Comment by ztelur** [comment]
I will fix the ci problem  soon

go: github.com/dubbogo/gost/remoting/etcd3 tested by
	github.com/dubbogo/gost/remoting/etcd3.test imports
	github.com/coreos/etcd/embed imports
	github.com/coreos/etcd/etcdserver imports
	github.com/coreos/etcd/mvcc/backend imports
	github.com/coreos/bbolt: github.com/coreos/bbolt@v1.3.5: parsing go.mod:
	module declares its path as: go.etcd.io/bbolt
	        but was required as: github.com/coreos/bbolt
Error: Process completed with exit code 1.

File: remoting/etcd3/client.go [hunk 1]
```diff
@@ -127,6 +127,8 @@ type Client struct {
 // NewClient create a client instance with name, endpoints etc.
 func NewClient(name string, endpoints []string, timeout time.Duration, heartbeat int) (*Client, error) {
 	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```
Comment (AlexStocks) [inline remoting/etcd3/client.go]: hey, guy, do u know what are you doing? If you invoke cancel,  the cancel will be useless on line 149.

**Comment by AlexStocks** [comment]
Pls ensure that the dir should be database/kv/etcd/v3 which is like go's dir 'https://github.com/golang/go/tree/master/src/database/sql'.

File: database/kv/etcd/v3/client.go [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+	// clean raw client
+	c.rawClient = nil
+}
+
+func (c *Client) stop() bool {
+	select {
+	case <-c.exit:
+		return true
+	default:
+		close(c.exit)
```
Comment (AlexStocks) [inline database/kv/etcd/v3/client.go:129]: using sync.Once

File: database/kv/etcd/v3/client.go [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+		return perrors.WithMessage(err, "new session with server")
+	}
+
+	// must add wg before go maintenance status goroutine
+	c.Wait.Add(1)
+	go c.maintenanceStatusLoop(s)
+	return nil
+}
+
+func (c *Client) maintenanceStatusLoop(s *concurrency.Session) {
```
Comment (AlexStocks) [inline database/kv/etcd/v3/client.go:166]: keepAlive?

File: database/kv/etcd/v3/client.go [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+			// when etcd server stopped, cancel ctx, stop all watchers
+			c.clean()
+			// when connection lose, stop client, trigger reconnect to etcd
+			c.stop()
+			c.lock.Unlock()
+			return
+		}
+	}
+}
+
```
Comment (AlexStocks) [inline database/kv/etcd/v3/client.go:189]: func (c *Client)GetClient() *etcdv3.Client {
	c.lock.RLock()
	defer c.lock.RUnlock()

	return c.rawClient
}

**Review by AlexStocks — APPROVED** [review]

**Review by gaoxinge — APPROVED** [review]

**Review by wongoo — APPROVED** [review]

**Pull request merged**